### ISSUE TITLE
Vert.x Cluster Corrupt When 2 or More Members are Shutdown Ungracefully

### ISSUE THREAD

**Issue by AlexThurston** [eclipse/vert.x#1737]
Vert.x Version: 3.3.3
Host: OSX

As the title suggests, it looks as though when 2 or more Vert.x instances are killed in an ungraceful manner (ie. SIGKILL instead of SIGINT/TERM), the list of active handlers becomes incorrect and the cluster is corrupt in such a way that is unrecoverable.  I'll start by providing a way to demonstrate the problem.

The way I have been reproducing this is by using the *pointtopoint* example provided in the Vert.x example (https://github.com/vert-x3/vertx-examples/tree/master/core-examples/src/main/java/io/vertx/example/core/eventbus/pointtopoint)

I edited the receiver slight to provide more information in the failure case.  It looks like this:
```java
import io.vertx.core.AbstractVerticle;
import io.vertx.core.eventbus.EventBus;
import io.vertx.core.json.JsonObject;

public class Receiver extends AbstractVerticle {
  @Override
  public void start() throws Exception {
    EventBus eb = vertx.eventBus();

    eb.consumer("ping-address", message -> {
      System.out.println("Received message: " + message.body());

      // Now send back reply
      JsonObject foo = new JsonObject();
      foo.put("foo","bar");
      message.reply(foo);
    });

    System.out.println("Receiver ready!");
  }
}
```

Similarly, I changed the Sender slightly:
```java
import io.vertx.core.AbstractVerticle;
import io.vertx.core.eventbus.EventBus;

public class Sender extends AbstractVerticle {
  @Override
  public void start() throws Exception {
    EventBus eb = vertx.eventBus();
    vertx.setPeriodic(1000, v -> {
      eb.send("ping-address", "ping!", reply -> {
        if (reply.succeeded()) {
          System.out.println("Received reply " + reply.result().body());
        } else {
          System.out.println("No reply: " + reply.cause());
        }
      });
    });
  }
}
```

In all the following experiments, I am running 4 Receivers:
`vertx run Receiver.java -cluster`

And a single Sender:
`vertx run Sender.java -cluster`

###Experiments
#### Gracefully stop 1 Receiver
In this case, I CTRL+C (or send a SIGINT) to a single Receiver.  In this case, the Sender continues to function normally and all is well

#### Gracefully stop 2+ Receivers
Using `kill <PID> <PID> <PID>` to kill multiple receivers gracefully _at the same time_.  In this case, all is well and the Sender continues to behave as expected.  You can see that the each Receiver is not stopped at the same time, but that upon receipt of the signal stops handling messages.  At no point does the Sender ever fail to have a message handled.

#### Ungracefully stop 1 Receiver
Using `kill -9 <PID>`, I kill one single Receiver.  Here you will notice that the Sender will receiver some timeouts in it's calls.  This presumably can be attributed to there being messages sent to the downed handler prior to the cluster recognizing that it is gone.  IMO, some failed calls in this case are probably expected as the cluster does recover once things have settled.

#### Ungracefully stop 2+ Receivers
Here is where the proverbial :poop: hits the fan.  In this case, I `kill -9 <PID> <PID> <PID>`.  At this point, it appears as though, undeterministically, what the remaining, clustered vertx process think remain versus what it is in the Hazelcast map is sometimes incorrect.  On occasion they are different and sometimes they are the same, but neither are ever correct.  You'll notice that the Sender begins to kill several calls.  This is because it still thinks that handlers that were just killed are still available.  Where it gets more interesting is that if I then check the contents of the Hazelcast map, it sometimes contains the now deleted handle endpoints, but on occasion will be entirely empty.  This means that without completely destroying and remaking the cluster, one cannot recover.  In a system that has many different services, this makes for some massive downtime.

It's worth noting that I do appreciate that kill -9 is a less than desirable thing to occur but the reality is that things do crash and that being able to recover from these crashes is crucial.

**Comment by AlexThurston** [comment]
I'll also add that the majority of this investigation was done using Vert.x 2.1.7 (which is what I am actually using).  I discovered that it was also a problem in Vert.x 3 and thought that because of that, it was more likely to be addressed :)

**Comment by bdarwin** [comment]
Yes I noticed this as well. The whole cluster becomes unstable even a single process gets killed.

**Comment by tsegismont** [comment]
Fixed in #1848
